src/Formik.tsx

```diff
@@ -259,92 +253,91 @@ export function useFormik<Values = object>({
-        return Promise.all([
-          runFieldLevelValidations(values),
-          props.validationSchema ? runValidationSchema(values) : {},
-          props.validate ? runValidateHandler(values) : {},
-        ]).then(([fieldErrors, schemaErrors, validateErrors]) => {
-          const combinedErrors = deepmerge.all<FormikErrors<Values>>(
-            [fieldErrors, schemaErrors, validateErrors],
-            { arrayMerge }
-          );
-          if (!isEqual(state.errors, combinedErrors)) {
-            dispatch({ type: 'SET_ERRORS', payload: combinedErrors });
-          }
-          dispatch({ type: 'SET_ISVALIDATING', payload: false });
-          return combinedErrors;
-        });
-      } else {
-        return Promise.resolve({});
-      }
+  // Are there any validation methods?
+  const hasValidationFns = React.useMemo(
+    () =>
+      props.validationSchema ||
+      props.validate ||
+      (fields.current &&
+        Object.keys(fields.current).filter(
+          key => !!fields.current[key].validate
+        ).length > 0),
+    [props.validate, props.validationSchema, fields]
+  );
+
+  // Run all validations and return the result
+  const runAllValidations = React.useCallback(
+    (values: Values) => {
+      return Promise.all([
+        runFieldLevelValidations(values),
+        props.validationSchema ? runValidationSchema(values) : {},
+        props.validate ? runValidateHandler(values) : {},
+      ]).then(([fieldErrors, schemaErrors, validateErrors]) => {
+        const combinedErrors = deepmerge.all<FormikErrors<Values>>(
+          [fieldErrors, schemaErrors, validateErrors],
+          { arrayMerge }
+        );
+        return combinedErrors;
+      });
     },
     [
       props.validate,
       props.validationSchema,
       runFieldLevelValidations,
       runValidateHandler,
       runValidationSchema,
-      state.errors,
-      state.values,
-      fields,
     ]
   );
 
-  React.useEffect(() => {
-    if (
-      prevState.values !== state.values &&
-      !!validateOnChange &&
-      !state.isSubmitting &&
-      isMounted.current != null
-    ) {
-      const [validate, cancel] = makeCancelable(validateForm());
-      validate.then(x => x).catch(x => x); // catch the rejection silently
-      return cancel;
-    }
-    return;
-  }, [
-    prevState.values,
-    state.isSubmitting,
-    state.values,
-    validateForm,
-    validateOnChange,
-    isMounted,
-  ]);
+  // Run validations and dispatching the result as low-priority via rAF.
+  //
+  // The thinking is that validation as a result of onChange and onBlur
+  // should never block user input. Note: This method should never be called
+  // during the submission phase because validation prior to submission
+  // is actaully high-priority since we absolutely need to guarantee the
+  // form is valid before executing props.onSubmit.
+  const validateFormWithLowPriority = React.useCallback(
```

Comment:
Makes sense

Reply:
FWIW the eslint-plugin-react-hooks gets mad when I remove the useCallback here. Thoughts?

Reply:
cc @gaearon. I usually just let the hooks plugin do its job. However, does @fredyc have a point about avoiding useCallback here?

Reply:
I don't use eslint plugin as it currently gets more in the way that it helps. Perhaps my understanding of `useCallback` is wrong right now, but cannot imagine why it would want to enforce it for such internal function.

Reply:
I think I am going to go through and pull out a lot of these functions in another PR. Let's keep this discussion focused on high/low pri.